src/app/utils/withStoresLoader.tsx

```diff
@@ -0,0 +1,80 @@
+import React, { useEffect } from 'react';
+import { t } from 'app/i18next-t';
+import { DestinyAccount } from 'app/accounts/destiny-account';
+import { D1StoresService } from 'app/inventory/d1-stores';
+import { D2StoresService } from 'app/inventory/d2-stores';
+import { DestinyComponentType, DestinyProfileResponse } from 'bungie-api-ts/destiny2';
+import { useSubscription } from './hooks';
+import { refresh$ } from 'app/shell/refresh';
+import { queueAction } from 'app/inventory/action-queue';
+import ShowPageLoading from 'app/dim-ui/ShowPageLoading';
+import { connect } from 'react-redux';
+import { RootState } from 'app/store/types';
+import { storesSelector, bucketsSelector, profileResponseSelector } from 'app/inventory/selectors';
+import { currentAccountSelector, destinyVersionSelector } from 'app/accounts/selectors';
+import type { DimStore } from 'app/inventory/store-types';
+import { InventoryBuckets } from 'app/inventory/inventory-buckets';
+import { D2ManifestDefinitions } from 'app/destiny2/d2-definitions';
+import { D1ManifestDefinitions } from 'app/destiny1/d1-definitions';
+import { DestinyVersion } from '@destinyitemmanager/dim-api-types';
+
+export interface StoresLoadedProps {
+  destinyVersion: DestinyVersion;
+  account: DestinyAccount;
+  profileInfo: DestinyProfileResponse;
+  defs: D2ManifestDefinitions | D1ManifestDefinitions;
+  defsD2: D2ManifestDefinitions;
+  stores: DimStore[];
+  buckets: InventoryBuckets;
+  isPhonePortrait: boolean;
+}
+
+function mapStateToProps(state: RootState): StoresLoadedProps {
+  const destinyVersion = destinyVersionSelector(state);
+
+  return {
+    destinyVersion,
+    account: currentAccountSelector(state)!,
+    profileInfo: profileResponseSelector(state)!,
+    defs: destinyVersion === 2 ? state.manifest.d2Manifest! : state.manifest.d1Manifest!,
+    defsD2: state.manifest.d2Manifest!,
+    stores: storesSelector(state),
+    buckets: bucketsSelector(state)!,
+    isPhonePortrait: state.shell.isPhonePortrait,
+  };
+}
+
+function getStoresService(account: DestinyAccount) {
+  return account.destinyVersion === 1 ? D1StoresService : D2StoresService;
+}
+
+/**
+ * Wraps a component and will preload account/profile/store information.
+ * Optionally provide components for a paired down request (otherwise fetches all)
+ */
+export default function withStoresLoader(WrappedComponent, components?: DestinyComponentType[]) {
+  function LoadedComponent(props: StoresLoadedProps & any) {
+    const { account, stores } = props;
+
+    useEffect(() => {
+      const storesService = getStoresService(account);
+      if (!stores.length) {
+        // TODO: Dispatch an action to load stores instead
+        storesService.getStoresStream(account, components);
```

Comment:
My worry here is that if you load with a stripped-down set of components, then try to go to a page that needs more, it won't trigger the update because stores is already loaded.

My plan was to move everything about stores loading into Redux (no more RxJS!) and then store a map from component type to the timestamp that component was last loaded. Each page can then provide which components it needs, and if the components it needs are up to date, no problem! Otherwise we load just the components that are out of date and merge into our cached profile.

Reply:
ah the additional components are loaded but it's async, so it should come in but wont block the initial load. yea the way you described sounds like the way to go

Reply:
Ah I see, I didn't quite follow that this was only for the very first load.